bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -3271,16 +3271,104 @@ private static boolean checkFillerValue(BUnionType type,  List<Type> unAnalyzedT
+        List<Type> nonFiniteTypes = new ArrayList<>();
+        Set<Object> allFiniteTypesValueSpace = new HashSet<>();
+        for (Type memberType: memberTypes) {
+            if (memberType.getTag() == TypeTags.FINITE_TYPE_TAG) {
+                allFiniteTypesValueSpace.addAll(((BFiniteType) memberType).getValueSpace());
+            } else {
+                nonFiniteTypes.add(memberType);
+            }
+        }
+
+        if (nonFiniteTypes.isEmpty()) {
+            // only finite types are there, so the check narrows to one finite type like case
+            return hasFillerValueInValueSpace(allFiniteTypesValueSpace);
+        } else {
+            // non finite types are available
+            Iterator<Type> iterator = nonFiniteTypes.iterator();
+            Type firstMember = iterator.next();
+
+            // non finite types are checked whether they are the same type
+            Type nextMember;
+            while (iterator.hasNext()) {
+                nextMember = iterator.next();
+                if (!isSameBasicType(firstMember, nextMember)) {
+                    return false;
+                }
+            }
+
+            // if no finite types the checking ends here
+            if (allFiniteTypesValueSpace.isEmpty()) {
+                return hasFillerValue(firstMember);
+            }
+
+            // both finite and non finite types are available
+            // finite types are checked whether they are the type of non finite types
+            if (!containsSameBasicType(firstMember, allFiniteTypesValueSpace)) {
+                return false;
+            }
+
+            // all members are same basic types
+            // need to check filler value is there
+            if (hasFillerValue(firstMember)) {
+                return true;
+            }
+            return (allFiniteTypesValueSpace.size() == 1) ?
+                    isFillerValueOfFiniteTypeBasicType(allFiniteTypesValueSpace.iterator().next()) :
+                    hasFillerValueInValueSpace(allFiniteTypesValueSpace);
+        }
+    }
+
+    private static boolean isSameBasicType(Type sourceType, Type targetType) {
+        if (isSameType(sourceType, targetType)) {
+            return true;
+        }
+        int sourceTag = sourceType.getTag();
+        int targetTag = targetType.getTag();
+        if (TypeTags.isStringTypeTag(sourceTag) && TypeTags.isStringTypeTag(targetTag)) {
+            return true;
+        }
+        if (TypeTags.isXMLTypeTag(sourceTag) && TypeTags.isXMLTypeTag(targetTag)) {
+            return true;
+        }
+        return isIntegerSubTypeTag(sourceTag) && isIntegerSubTypeTag(targetTag);
+    }
+
+    private static boolean isIntegerSubTypeTag(int typeTag) {
+        return TypeTags.isIntegerTypeTag(typeTag) || typeTag == TypeTags.BYTE_TAG;
+    }
+
+    private static boolean isFillerValueOfFiniteTypeBasicType(Object value) {
+        if (value == null) {
```

Comment:
Shall we add a test to cover this line? with the `()` value as a singleton type.